packages/server/src/Connection.ts

```diff
@@ -41,6 +41,10 @@ export class Connection {
 
   logger: Debugger
 
+  private beforeHandleMessageQueue: Uint8Array[] = []
```

Comment:
You might be able to do with just 1 additional state variable if you unshifted the message only after successfully processing it.